File: python/stripchart.py
#
# stripchart
#    
#    implementation ot fhe <stripChart> class. This class updates the screen like a old fashioned 
#    paper recorder, old data is shifted to the left and new data is added on the right 
#
# modifications
#    26-jan-2024  JM   initial version
#    27-feb-2024  JM   now derived from graphChart

from graphchart import graphChart
from PySide6.QtCore import QPointF

class stripChart(graphChart) :

  # constructor

  def __init__(self,chanlist) :
    super().__init__(chanlist)
    
  # updatePlot
  #
  #   updates the graph with new samples for one channel

  def updatePlot(self,ichan,data) :

    # to make it faster

    indx = self.m_indx[ichan]     
    deltaT = self.m_deltaT[ichan]
    maxindx = self.m_pntsInGraph[ichan]
    
    # downSample 

    data = self.m_downSampler[ichan].getData(data)   
    nsamples = len(data)    

    if (self.m_firstScreen == True) :

      # first screen, points should be appended to buffer. However if the end is reached already 
      # existing points should be shifted
  
      shift = nsamples - (maxindx - indx)  
      if (shift > 0) :
        
        for i in range(shift,indx) : self.m_dataBuffer[ichan][i-shift] = QPointF((i-shift) * deltaT,self.m_dataBuffer[ichan][i].y())
        for i,j in zip(range(indx-shift,indx),range(shift)) : self.m_dataBuffer[ichan][i] = QPointF(i * deltaT,data[j])
        for i in range(nsamples-shift) :
          self.m_dataBuffer[ichan].append(QPointF(indx * deltaT,data[i+shift]))
          indx += 1
     
      else :

        for i in range(nsamples):
          self.m_dataBuffer[ichan].append(QPointF(indx * deltaT,data[i]))
          indx += 1

    else :
      
      # other screens, only a shift is required and new samples should be added
            
      shift = maxindx - nsamples
      for i in range(shift) :
        self.m_dataBuffer[ichan][i] = QPointF((i * deltaT),self.m_dataBuffer[ichan][i+nsamples].y()) 
      for i in range(shift,maxindx) :
        self.m_dataBuffer[ichan][i] = QPointF(i * deltaT,data[i-shift])
    
    # done, update index

    self.m_indx[ichan] = indx

File: matlab/stripchart.m
%
% stripchart
%
%     this is the implementation of stripchart, a routine to plot signals like it is done
%     on a traditional paper recorder. Old signals are removed from the left of the screen
%     and new signals are added to the right while the screen is shifted
%
% modifications
%     03-feb-2024   JM    initial version

classdef stripchart < basechart

  methods

    %% constructor

    function obj = stripchart(axisHandle,channels)

      % in the constructor the constructor of the parent class is called and some
      % variables are initialised
      %
      %     syntax : obj = stripchart(axisHandle,channels)
      %
      % with <axixHandle> a handle to the axis of the chart and <channels> a list of
      % channels that should be plotted in this chart

      obj = obj@basechart(axisHandle,channels);

    end


    %% update

    function update(obj,ichan,data)

      % <update> updates the graph for channel <ichan> with the data from <data>. Old
      % points are removed on left the graph is shifted and the new data points are added
      % to the right
      %
      %     syntax : update(obj,ichan,data)
      %
      % with <ichan> the channel that should be updated and <data> the new data that is
      % added to the graph
      
      indx = obj.m_indx(ichan);
      maxindx = obj.m_pntsInGraph(ichan);
      
      % downsample

      data = obj.m_downSampler{ichan}.getData(data);
      nsamples = length(data);
      
      % in the first screen points are added, to an empty buffer, later they are shifted
      
      if (obj.m_firstScreen)

        shift = indx + nsamples -maxindx;
        if shift > 0
          obj.m_dataBuffer(ichan).y = [obj.m_dataBuffer(ichan).y((shift+1):end) data];
        else
          obj.m_dataBuffer(ichan).y = [obj.m_dataBuffer(ichan).y data];
        end

      else    
        obj.m_dataBuffer(ichan).y = [obj.m_dataBuffer(ichan).y((nsamples+1):end) data];
      end
      
      % done, store current position

      obj.m_indx(ichan) = length(obj.m_dataBuffer(ichan).y);

    end

  end
end
